File: Data Structures & Algorithms/target-sum/submission-2.cpp
class Solution
{
public:
    int findTargetSumWays(vector<int> &nums, int target)
    {
        int sum = 0;
        for (int num : nums)
        {
            sum += num;
        }

        vector<vector<int>> dp(nums.size() + 1, vector<int>(2 * sum + 1, -1));
        return dfs(0, 0, nums, target, dp, sum);
    }

private:
    int dfs(int index, int total, vector<int> &nums, int target, vector<vector<int>> &dp, int sum)
    {
        int temp = total + sum;
        if (index == nums.size() && total == target)
        {
            dp[index][temp] = 1;
            return 1;
        }
        else if (index == nums.size())
        {
            dp[index][temp] = 0;
            return 0;
        }
        if (dp[index][temp] != -1)
        {
            return dp[index][temp];
        }

        int ans = dfs(index + 1, total - nums[index], nums, target, dp, sum) + dfs(index + 1, total + nums[index], nums, target, dp, sum);
        dp[index][temp] = ans;
        return ans;
    }
};

File: Data Structures & Algorithms/interleaving-string/submission-2.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        memo = {}
        def dfs(i, j, k):
            if k == len(s3):
                return True if i == len(s1) and j == len(s2) else False
            if j > len(s2):
                return False
            if i > len(s1):
                return False
            if (i, j, k) in memo:
                return memo[(i, j, k)]
            ans = False
            if i < len(s1) and s1[i] == s3[k]:
                ans = ans or dfs(i + 1, j, k + 1)
            if j < len(s2) and s2[j] == s3[k]:
                ans = ans or dfs(i, j + 1, k + 1)
            memo[(i, j, k)] = ans
            return ans
        return dfs(0, 0, 0)